### PR TITLE
Use Builders for `array_data_from_js`

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

For loading the arrow data from JavaScript, we previously did some real low level stuff, which is not needed anymore.

## 🐾 Next steps

Use more builders in the rest of the code

## 🔍 What does this change?

- Changes `array_data_from_js` to use an `ArrayDataBuilder` instead of building int by hand
- Use builders for data where possible

It's not possible (or not a good idea) to use non-buffer builders, as the data already exists (behind pointers). Otherwise the data would need to be parsed and recreated from scratch, which would hit the performance badly.

### 📜 Does this require a change to the docs?

No

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1201793759102797/1201877648275953/f) (_internal_)

## 🛡 What tests cover this?

None directly, this function is indirectly tested by integration tests.

### CODE REVIEW

File: packages/engine/src/worker/runner/javascript/mod.rs
Comment (Alfred-Mountfield) [inline packages/engine/src/worker/runner/javascript/mod.rs:568]: I'm guessing the unsafe is to do with the `slice::from_raw_parts`? Perhaps we could add a safety comment above the match or something to describe why it's safe. 

Out of curiosity why have you also put the builder code within the unsafe, are their operations unsafe if the values array is borked?

File: packages/engine/src/worker/runner/javascript/mod.rs
Comment (Alfred-Mountfield) [inline packages/engine/src/worker/runner/javascript/mod.rs:593]: The lack of explanation in this file is worrying to me. Especially the inconsistency with `target_len` and `target_len + 1` and also the TODO above.

I appreciate you migrated this code and did not write most of it, so perhaps it's unrealistic to expect you to be able to do so, but can you figure it out and document it better? 

If it's a lot of work then lets add a TODO to improve the readability and explanation in this code. I specifically want to know at a glace why we're using target_len instead of data.len (e.g. testing the TODO above which we might need to wait until tests pass and a merge with main), and also why we're adding +1 to the len sometimes, etc.

Comment (Alfred-Mountfield) [inline packages/engine/src/worker/runner/javascript/mod.rs:593]: The lack of explanation in this file is worrying to me. Especially the inconsistency with `target_len` and `target_len + 1` and also the TODO above.

I appreciate you migrated this code and did not write most of it, so perhaps it's unrealistic to expect you to be able to do so, but can you figure it out and document it better? 

If it's a lot of work then lets add a TODO to improve the readability and explanation in this code. I specifically want to know at a glace why we're using target_len instead of data.len (e.g. testing the TODO above which we might need to wait until tests pass and a merge with main), and also why we're adding +1 to the len sometimes, etc.

File: packages/engine/src/worker/runner/javascript/mod.rs
Comment (Alfred-Mountfield) [inline packages/engine/src/worker/runner/javascript/mod.rs:563]: Is this TODO still valid? I don't quite understand it but you have seemed to change it since the original code

File: packages/engine/src/worker/runner/javascript/mod.rs
Comment (Alfred-Mountfield) [inline packages/engine/src/worker/runner/javascript/mod.rs:642]: See it also confuses me here as to why we're using `data.len` instead of `target_len`

File: packages/engine/src/worker/runner/javascript/mod.rs
Comment (TimDiekmann) [reply packages/engine/src/worker/runner/javascript/mod.rs:582]: Yes, we are not using the original pointer from JavaScript, thus we copying the data. Previously this was done by `.into()` (one reason I don't like the trait, it's hard to see clones), now it's done by `Buffer::from_slice_ref`.

File: packages/engine/src/worker/runner/javascript/mod.rs
Comment (Alfred-Mountfield) [inline packages/engine/src/worker/runner/javascript/mod.rs:658]:
```suggestion
                let values = slice::from_raw_parts(data.buffer_ptrs[0], *size as usize * data.len);
```
Slightly easier to parse

File: packages/engine/src/worker/runner/javascript/mod.rs
Comment (Alfred-Mountfield) [inline packages/engine/src/worker/runner/javascript/mod.rs:667]: Inconsistent with the unsafe blocks above in terms of how much is inside its scope

File: packages/engine/src/worker/runner/javascript/mod.rs
Comment (Alfred-Mountfield) [reply packages/engine/src/worker/runner/javascript/mod.rs:582]: Can we rephrase the TODO then. So the TODO isn't extra copies, it's to _stop making_ extra copies?

Comment (Alfred-Mountfield) [reply packages/engine/src/worker/runner/javascript/mod.rs:582]: Also the reasoning you've said isn't explained in the TODO so it would be good to do so.

File: packages/engine/src/worker/runner/javascript/mod.rs
Comment (Alfred-Mountfield) [inline packages/engine/src/worker/runner/javascript/mod.rs:582]:
```suggestion
        // TODO: We currently copy the buffers because the JavaScript representation of arrays does
```

Can you say how we copy the buffers. As in what's causing the copy (I believe you said it's `Buffer::from_slice_ref`?

File: packages/engine/src/worker/runner/javascript/mod.rs
Comment (Alfred-Mountfield) [inline packages/engine/src/worker/runner/javascript/mod.rs:632]:
```suggestion
                // the previous value, thus we need `num_values + 1` offset values.
                
```

File: packages/engine/src/worker/runner/javascript/mod.rs
Comment (Alfred-Mountfield) [inline packages/engine/src/worker/runner/javascript/mod.rs:633]: Can we reorder these comments and add a line-break between the bits, like the TODO and the actual comment and also the SAFETY point

**Review by Alfred-Mountfield — CHANGES_REQUESTED** [review]

File: packages/engine/src/worker/runner/javascript/mod.rs
Comment (Alfred-Mountfield) [inline packages/engine/src/worker/runner/javascript/mod.rs:628]: Do we want to get rid of this TODO? You had `data.len` in some places and it caused errors didn't it?

File: packages/engine/src/worker/runner/javascript/mod.rs
Comment (TimDiekmann) [inline packages/engine/src/worker/runner/javascript/mod.rs:635]:
```suggestion
                //   (In practice, target_len has worked for a long time, though that's not an ideal
                //   reason to use it. Maybe `data.len` would also work.)
                
                // The offsets are in the first buffer. For each value in the second buffer, we have
                // a start offset and an end offset. The start offset is equal to the end offset of
                // the previous value, thus we need `num_values + 1` offset values.
                
                // SAFETY: `data.buffer_ptrs[0]` is provided by arrow as offsets, the type is `i32`,
                //   and `target_len + 1` is carefully chosen.
```

File: packages/engine/src/worker/runner/javascript/mod.rs
Comment (Alfred-Mountfield) [inline packages/engine/src/worker/runner/javascript/mod.rs:679]:
```suggestion
                    Some(*size as usize * target_len),
```
is a little more readable

File: packages/engine/src/worker/runner/javascript/mod.rs
Comment (Alfred-Mountfield) [inline packages/engine/src/worker/runner/javascript/mod.rs:703]:
```suggestion
                //   `*size as usize * data.len` corresponds for the `size` of *each* value.
                let values = unsafe {
                    slice::from_raw_parts(data.buffer_ptrs[0], *size as usize * data.len)
```

Slightly more readable

File: packages/engine/src/worker/runner/javascript/mod.rs
Comment (TimDiekmann) [reply packages/engine/src/worker/runner/javascript/mod.rs:632]: I added a newline between the comment and the SAFETY section. The Todo was removed as we don't want to use `data.len` anymore.

File: packages/engine/src/worker/runner/javascript/mod.rs
Comment (TimDiekmann) [inline packages/engine/src/worker/runner/javascript/mod.rs:699]:
```suggestion
                //   `*size as usize * target_len` corresponds for the `size` of *each* value.
```

File: packages/engine/src/worker/runner/javascript/mod.rs
Comment (Alfred-Mountfield) [inline packages/engine/src/worker/runner/javascript/mod.rs:701]:
```suggestion
                    slice::from_raw_parts(data.buffer_ptrs[0], *size as usize * target_len)
```

**Review by Alfred-Mountfield — CHANGES_REQUESTED** [review]

File: packages/engine/src/worker/runner/javascript/mod.rs
Comment (Alfred-Mountfield) [inline packages/engine/src/worker/runner/javascript/mod.rs:583]:
```suggestion
        //   JavaScript representation of arrays does not match the Rust implementation. Try to
```

**Review by Alfred-Mountfield — APPROVED** [review]

**Pull request merged**
